src/tests/rules/valid-repository.test.ts

```diff
@@ -0,0 +1,211 @@
+						error: `repository is missing property "url", which should be the url to a repository (e.g. "git+https://github.com/npm/cli.git")`,
+					},
+					line: 2,
+					messageId: "validationError",
+				},
+			],
+		},
+		{
+			code: `{
+	"repository": {
+        "": "packages/lib-a",
+        "  ": "git",
+        "    ": "git+https://github.com/JoshuaKGoldberg/package-json-validator.git"
+    }
+}
+`,
+			errors: [
+				{
+					data: {
+						error: `repository is missing property "type", which should be the type of repository this is (e.g. "git")`,
+					},
+					line: 2,
+					messageId: "validationError",
+				},
+				{
+					data: {
+						error: `repository is missing property "url", which should be the url to a repository (e.g. "git+https://github.com/npm/cli.git")`,
+					},
+					line: 2,
+					messageId: "validationError",
+				},
+				{
+					data: {
+						error: 'property 0 is invalid; keys should be "type", "url", and optionally "directory"',
+					},
+					line: 3,
+					messageId: "validationError",
+				},
+				{
+					data: {
+						error: 'property 1 is invalid; keys should be "type", "url", and optionally "directory"',
+					},
+					line: 4,
+					messageId: "validationError",
+				},
+				{
+					data: {
+						error: 'property 2 is invalid; keys should be "type", "url", and optionally "directory"',
+					},
+					line: 5,
+					messageId: "validationError",
+				},
+			],
+		},
+		...[
+			"svn:npm/example",
+			"eslint-plugin-package-json",
+			"git:npm/example",
+			"github:npm/example/repo",
+			"org/user/repo",
+		].map((value) => ({
+			code: `{
+	"repository": "${value}"
+}
+`,
+			errors: [
+				{
+					data: {
+						error: `the value "${value}" is invalid; it should be the shorthand for a repository (e.g. "github:npm/example")`,
+					},
+					line: 2,
+					messageId: "validationError",
+				},
+			],
+		})),
+	],
+	valid: [
+		"{}",
+		...[
+			"git+https://github.com/JoshuaKGoldberg/package-json-validator.git",
+			"https://github.com/JoshuaKGoldberg/package-json-validator",
+			"https://github.com/JoshuaKGoldberg/package-json-validator.git",
+			"http://github.com/JoshuaKGoldberg/package-json-validator.git",
+			"git://github.com/JoshuaKGoldberg/package-json-validator.git",
+			"git://github.com/JoshuaKGoldberg/package-json-validator",
+			"git@github.com:JoshuaKGoldberg/package-json-validator.git",
+		]
+			.map((value) => [
+				{
+					code: `{
+	"repository": {
+      "type": "git",
+      "url": "${value}",
+      "directory": "packages/a"
+    }
+}
+`,
+					name: `${value} (with directory)`,
+				},
+				{
+					code: `{
+	"repository": {
+      "type": "git",
+      "url": "${value}"
+    }
+}
+`,
+					name: `${value} (without directory)`,
+				},
+			])
+			.flat(),
+		...[
+			"npm/example",
+			"github:npm/example",
+			"gist:11081aaa281",
+			"bitbucket:user/repo",
+			"gitlab:user/repo",
+		].map((value) => ({
+			code: `{
+	"repository": "${value}"
+}
+`,
+			name: value,
+		})),
+	],
```

Comment:
I'm not a huge fan of auto-generating tests. I've advocated for banning them altogether in typescript-eslint, and will eventually try to spread the practice elsewhere: https://github.com/typescript-eslint/typescript-eslint/issues/11052. But not a blocker IMO. This repo is a lot simpler than typescript-eslint's.

Reply:
Hmm, I'm not sure I fully agree in this case.  I think, like with anything, it can be taken too far, and result in hard to reason about tests.  For instance, I don't at all like the way many of the test suites in `eslint-plugin-import` are constructed (e.g. https://github.com/import-js/eslint-plugin-import/blob/main/tests/src/rules/no-cycle.js#L287-L311).  I've always found those harder to work with and a bit unapproachable.  

But in simple cases like this where the test code is all going to be exactly the same except for one prop value, it's an easy way to reduce some boilerplate.  I guess what I'm trying to say is that it doesn't feel like quite so black and white to me, and extremes are bad for most things 😄